src/main/scala/org/opennetworkinsight/netflow/FlowPostLDA.scala

```diff
@@ -63,67 +65,51 @@ object FlowPostLDA {
 
     logger.info("Computing conditional probability")
 
-    val dataWithSrcScore = score(sc, dataWithWord, ipToTopicMixes, wordToProbPerTopic, SourceScore, SourceIP, SourceWord)
-    val dataWithDestScore = score(sc, dataWithSrcScore, ipToTopicMixes, wordToProbPerTopic, DestinationScore, DestinationIP, DestinationWord)
-    val dataScored = minimumScore(dataWithDestScore)
-
-    logger.info("Persisting data")
-    val filteredDF = dataScored.filter(MinimumScore + " <" + threshold)
-
-    val count = filteredDF.count
+    val docToTopicMixRDD: RDD[(String, Array[Double])] = sc.parallelize(docToTopicMix.toSeq)
 
-    val takeCount  = if (topK == -1 || count < topK) {
-      count.toInt
-    } else {
-      topK
-    }
-
-    val minimumScoreIndex = filteredDF.schema.fieldNames.indexOf(MinimumScore)
-
-    class DataOrdering() extends Ordering[Row] {
-      def compare(row1: Row, row2: Row) = row1.getDouble(minimumScoreIndex).compare(row2.getDouble(minimumScoreIndex))
-    }
+    val docToTopicMixDF = docToTopicMixRDD.map({ case (doc, probabilities) => DocTopicMix(doc, probabilities) }).toDF
 
-    implicit val ordering = new DataOrdering()
-
-    val top: Array[Row] = filteredDF.rdd.takeOrdered(takeCount)
+    val words = sc.broadcast(wordToProbPerTopic)
 
-    val outputRDD = sc.parallelize(top).sortBy(row => row.getDouble(minimumScoreIndex))
+    val dataWithSrcScore = score(sc, dataWithWord, docToTopicMixDF, words, SourceScore, SourceIP, SourceProbabilities, SourceWord)
+    val dataWithDestScore = score(sc, dataWithSrcScore, docToTopicMixDF, words, DestinationScore, DestinationIP, DestinationProbabilities, DestinationWord)
+    val dataScored = minimumScore(dataWithDestScore)
 
-    // Using dropRight as we don't need last column MinimumScore, only SourceScore and DestinationScore
-    outputRDD.map(row => Row.fromSeq(row.toSeq.dropRight(1))).map(_.mkString(outputDelimiter)).saveAsTextFile(resultsFilePath)
+    logger.info("Persisting data")
+    val filteredDF = dataScored.filter(MinimumScore + " < " + threshold)
+    filteredDF.orderBy(MinimumScore).limit(topK).rdd.map(row => Row.fromSeq(row.toSeq.dropRight(1))).map(_.mkString(outputDelimiter)).saveAsTextFile(resultsFilePath)
```

Comment:
i honestly don't know... it comes down to whether the catalyst optimizer is smart enough to combine the orderBy and the limit to form a takeOrdered ... otherwise it might sort the entire data frame and then take the first topK elements....